lib/ModuleLayerCache.js

```diff
@@ -0,0 +1,81 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author @mknichel
+*/
+
+"use strict";
+
+/**
+ * A class used to represent a key in a {@link WeakMap}. {@link Symbol} can not
+ * be used due to https://github.com/nodejs/node/issues/49135.
+ */
+class WeakMapKey {
+	/**
+	 * @param {string} key The key to use for caching.
+	 */
+	constructor(key) {
+		this.key = key;
+	}
+}
+
+/**
+ * A cache for storing information that could be shared by modules across
+ * different layers. This can be useful for memory management since the same
+ * file that appears in multiple layers may duplicate large strings, such as
+ * the contents of the source file. The values in this cache can be garbage
+ * collected if there are no more retaining references.
+ */
+class ModuleLayerCache {
+	constructor() {
+		/** @type {Map<string, WeakMapKey>} */
+		this.weakMapKeys = new Map();
+		/** @type {WeakMap<WeakMapKey, string | Buffer>} */
+		this.cachedValues = new WeakMap();
```

Comment:
This is not "weak". `this.weakMapKeys` holds a strong reference to the `WeakMapKey`, so it can't be collected while `ModuleLayerCache` is active. So the whole class is equal to a single `Map<string, string | Buffer>`

Reply:
Is there another key that you could use? If so, just let me know. You need to use `userRequest` (a string) and therefore it has to be able to be a key in the `WeakMap` somehow. This holds on to the values during the lifetime of the `Compilation` object but then is thrown away.

Reply:
I think you can't make it "weak" on the values. That's won't work with string keys. But I think that might not be needed. But you need to make it "weak" on `associatedObjectForCache` that should be enough.